File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {

        vector<char> answer;

        int length = s.length();


        for(int i = 0; i < length; i++){
            switch(s[i]){
                case '(':
                    answer.push_back(s[i]);
                    break;
                case ')':
                    if(answer.size() == 0){
                        return false;
                    }
                    if(answer.back() == '('){
                        answer.pop_back();
                    }
                    else{
                        return false;
                    }
                    break;
                case '{':
                    answer.push_back(s[i]);
                    break;
                case '}':
                    if(answer.size() == 0){
                        return false;
                    }
                    if(answer.back() == '{'){
                        answer.pop_back();
                    }
                    else{
                        return false;
                    }
                    break;
                case '[':
                    answer.push_back(s[i]);
                    break;
                case ']':
                    if(answer.size() == 0){
                        return false;
                    }
                    if(answer.back() == '['){
                        answer.pop_back();
                    }
                    else{
                        return false;
                    }
                    break;
                default:
                    break;
            }
        }
        if(answer.size() == 0){
            return true;
        }
        return false;
        
    }
};

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = []
        open_brace = {'(':')', '{':'}','[':']'}
        close_brace = {')':'(', '}':'{',']':'['}


        # this will work if we think the parenthesis of 
        # seen = {'(':0, '{':0,'[':0}
        # for c in s:
        #     if c in open_brace:
        #         seen[c] += 1
        #     elif c in close_brace:
        #         seen[close_brace[c]] -= 1
        #         if seen[close_brace[c]] < 0:
        #             print(seen[close_brace[c]])
        #             return False
        

        # return True

        stack = []
        braces = {')': '(', '}': '{', ']': '['}
        
        for char in s:
            if char in braces:
                if stack:
                    top = stack.pop() 
                else:
                    top = None
                if braces[char] != top:
                    return False
            else:
                stack.append(char)
        
        return not stack

        